Src/IronPython/Modules/_bytesio.cs

```diff
@@ -146,59 +145,39 @@ public BytesIO(CodeContext/*!*/ context, object initial_bytes=null)
             [Documentation("readinto(array_or_bytearray) -> int.  Read up to len(b) bytes into b.\n\n"
                 + "Returns number of bytes read (0 for EOF)."
                 )]
-            public BigInteger readinto([NotNull]ByteArray buffer) {
-                _checkClosed();
-
-                int len = Math.Min(_length - _pos, buffer.Count);
-                for (int i = 0; i < len; i++) {
-                    buffer[i] = _data[_pos++];
-                }
-
-                return len;
-            }
-
-            public BigInteger readinto([NotNull]ArrayModule.array buffer) {
-                _checkClosed();
+            public BigInteger readinto([NotNull] IBufferProtocol buffer) {
+                try {
+                    using var pythonBuffer = buffer.GetBuffer(BufferFlags.Writable);
+                    var span = pythonBuffer.AsSpan();
 
-                int len = Math.Min(_length - _pos, buffer.__len__() * buffer.itemsize);
-                int tailLen = len % buffer.itemsize;
-                buffer.FromStream(new MemoryStream(_data, _pos, len - tailLen, false, false), 0);
-                _pos += len - tailLen;
+                    _checkClosed();
 
-                if (tailLen != 0) {
-                    byte[] tail = buffer.RawGetItem(len / buffer.itemsize);
-                    for (int i = 0; i < tailLen; i++) {
-                        tail[i] = _data[_pos++];
+                    int len = Math.Min(_length - _pos, span.Length);
+                    for (int i = 0; i < len; i++) {
+                        span[i] = _data[_pos++];
                     }
-                    buffer.FromStream(new MemoryStream(tail), len / buffer.itemsize);
-                }
 
-                return len;
+                    return len;
+                } catch (BufferException) {
+                    throw PythonOps.TypeError("readinto() argument must be read-write bytes-like object, not {0}", PythonTypeOps.GetName(buffer));
+                }
             }
 
             public override BigInteger readinto(CodeContext/*!*/ context, object buf) {
-                if (buf is ByteArray bytes) {
-                    return readinto(bytes);
-                }
-
-                if (buf is ArrayModule.array array) {
-                    return readinto(array);
-                }
-
-                _checkClosed();
-                throw PythonOps.TypeError("must be read-write buffer, not {0}", PythonTypeOps.GetName(buf));
+                var bufferProtocol = Converter.Convert<IBufferProtocol>(buf);
```

Comment:
Not ideal since it throws the wrong error message.

Reply:
I think the message is close enough. But for what types this overload is intended to work?

Reply:
Honestly I don't know, but it's an override of a virtual method so I don't know if it'll end up getting invoked somewhere via C#?

Reply:
OK, then even more reasons to believe that the message is good enough...